[utils] Add --{cc,cxx} options to build-script

vedantk opened the pull request:
Make it a bit easier to use a custom compiler or shim script to build swift & co.

Address review comments: drop the "with" prefix, only grab `host_clang` when it's needed.

This is an updated version of an earlier PR: https://github.com/apple/swift/pull/1037#issuecomment-173693840. I should have reused the PR but accidentally deleted my branch.

vedantk commented:
Regarding Jordan's comment about inferring cc from cxx or vice versa: would we have to ask the user if the build-script's guess is ok?

gribozavr commented:
I'd suggest that we don't try to infer, since it is not obvious where to insert the '++' if the 'cc' name is not 'clang'.  Some distributions append version numbers in different formats, or platform names, or variant names or add prefixes.

gribozavr commented on utils/build-script:
```diff
@@ -662,6 +670,20 @@ the number of parallel build jobs to use""",
+            return 1
+
+        if not args.cc:
+            args.cc = host_clang.cc
```
Instead of modifying command line parameters, could you just re-use the `host_clang` variable?  You can create an instance with `CompilerExecutable(cc=theCC, cxx=theCXX)`.

gribozavr commented on utils/build-script:
```diff
@@ -514,6 +514,14 @@ name of the directory under $SWIFT_BUILD_ROOT where the build products will be
 
+    parser.add_argument("--cc",
+        help="specify the C compiler to be used",
+        action="store", dest="cc")
```
Do we need the `dest=` here?  It seems like it can be inferred, like we do for for `--ios`.

jrose-apple commented:
I would certainly *allow* overriding, but in the simple case where we have just --cc passed, and doing one of the following transformations resolves to an executable, I'd prefer to have that executable chosen.

CC | CXX
--- | ---
gcc | g++
cc | c++
clang | clang++

If not that, can we at least give a warning for providing one but not the other? That's probably never intended.

vedantk commented:
Addressed review comments.

This version uses CompilerExecutable to avoid constructing `host_clang` twice, and to provide 'smart' suggestions when either CC or CXX are missing. It's done in a longest-suffix-first way which should address Dmitri's concern.

jrose-apple commented:
Ah, I was suggesting also checking that there is an executable file on disk by that name. Otherwise, it's probably an error.

vedantk commented:
Sure, but IMO that's what configure is for.

gribozavr commented on utils/build-script:
```diff
@@ -40,6 +40,47 @@ import swift_build_support.cmake
+    def _find_companion_compiler(self, cmd):
+        suggestions = [
+            ("cc", "c++"),
+            ("gcc", "g++"),
```
We don't support building Swift with gcc.

gribozavr commented on utils/build-script:
```diff
@@ -40,6 +40,47 @@ import swift_build_support.cmake
 
 
+# Utility for selecting host compilers.
+class CompilerExecutable(object):
```
Why not use the existing `CompilerExecutable` type instead, and turn the constructor into a global function?  Keeping the number of types low is good, IMHO.

gribozavr commented on utils/build-script:
```diff
@@ -662,6 +711,12 @@ the number of parallel build jobs to use""",
         ]
 
+    try:
+        compiler = CompilerExecutable(args)
```
Let's call the variable `host_compiler`.  I know other contributors are working on cross-compilation support.

gribozavr commented:
@jrose-apple CMake will fail with a good enough error if the compiler does not exist.

vedantk commented:
After thinking about this a bit more, I don't think build-script should make guesses about which host compiler the user wants. It's a bit messy and a minor convenience at best.

Stepping back even further, ISTM the best way to get this functionality is to append "-- --host-cc CC --host-cxx CXX" to the build-script invocation. Maybe we can revisit this with SR-237 later.
